### ISSUE TITLE
JSONDataFilter always mutates Map

### ISSUE THREAD

**Issue by gregw** [cometd/cometd#762]
The JSONDataFilter always calls setValue on map entries, even if the value has not been mutated.  This causes `UnsupportedOperationException` if a immutable map is passed even if no mutations are done.

**Comment by Mlong2177** [comment]
Mutating the passed object is unexpected behavior, in and of itself.

Suppose the following psuedo code.

```
private final static Map<String, Object> helloMessage = new HashMap<>();

//Code to  initialize map with the following:
//{"message": "Hello & Welcome, someone entered the channel."};

@Configure(value = "/hellochat")
public void configure(ConfigurableServerChannel channel){
	channel.addListener( new HTMLAmpersandEscaper() );
	channel.addListener( new ServerChannel.subscriptionListener() {
		public void subscribed(ServerSession session, ServerChannel channel, ServerMessage serverMessage) {
    			channel.publish(session, helloMessage);
		}
		//empty unsubscribe message here
	}
}


public class HTMLAmpersandEscaper extends JSONDataFilter {
	public String filterString(String string){
		return Pattern.compile("&").matcher(string).replaceall("&amp;");
	}
}
```

Wouldn't this lead to the escaper having run on the same text multiple times if multiple people subscribe to /hellochat?  That is very unintuitive behavior.

It seems to me that the solution should be cloning the map into a mutable version, and mutating the clone.

**Comment by gregw** [comment]
While the DataFilter interface does have the capability of cloning if so desired, the filter is applied to the Mutable message.  I think the design is such that cometd switched to the immutable version before sending to multiple subscribers.... but we'll review... @sbordet what do you think

**Comment by sbordet** [comment]
This is a case of a broadcast channel receiving a message and republishing the message _on the same channel_, which is wrong (the message will be published twice).

There is no need to republish the message, since the default Comet behavior for messages received on a broadcast channel is to publish to all subscribers.

**Comment by Mlong2177** [comment]
Except the re-publishing a message multiple times is intentional and required.  Further, you'll notice in the example I gave, it was NOT in response to receiving a message, but rather in response to the subscription to a channel.

**Comment by gregw** [comment]
I reopened because looking at JSONDataFilter, it is very inconsistent and inefficient about how it mutates data structures.  It always replaces Lists and Collections with new arrays, it modifies arrays directly and attempts to modify Maps (even if immutable).

ac11379 is a branch where I have made this behaviour consistent.  New containers are created only if the filter mutates a value and a new container is then always created
